### DISCUSSION TITLE
What are the prerequsites for new quests in SCEE? / How to go about suggesting new quests for SCEE?

### DISCUSSION THREAD

**Discussion by wielandb** [Helium314/SCEE#459] [General]
Hey everyone,

I have a lot of ideas for quests for SCEE that wouldn't be the right fit for StreetComplete for a variety of reasons. I would like to see theese quests in SCEE and would even be ready to happily implement most of them myself, but it got me wondering if, even if the quest guidelines are much less strict, which do still apply here?

Let's review the rules for StreetComplete:

- To be added tag is established and has a useful purpose
  - I guess that a tag would still have to have a defined meaning, but does "useful purpose" apply to SCEE? Can a quest exist that adds a tag that has a defined meaning, but where no major app exists that uses it (yet)? 
- Any answer the user can give must have an equivalent tagging (Quest should not reappear to other users when solved by one)
  - Does this apply to SCEE? iirc, quests in SCEE can be designed around "Sometimes, there is no answer and the user should just select "hide"" 
- Easily answerable by any pedestrian from the outside but a survey is necessary
  - I guess this applys, maybe minus the "easily by any pedestrian" part 
- Not an overwhelming percentage of quests have the same answer (No spam)
  - Does this apply? 
- Applies to a reasonable number of map data (Worth the effort)
  - I think it applys, but would it still if I would implement the quests myself?

My second, intertwined, question is how I would go about "proposing" a large number of quests. Should I open an issue for every single one and state that I would implement it myself? Should I just open pull requests? Should I open a discussion thread for all my ideas (one for every single one)? Should I start by opening and implementing one, and only after that is done another?

**Comment by mnalis** [comment]
Good questions, @wielandb! Here is my opinion (not particularly hard, i.e. I'm quite open to discussion about pros/cons)

> - To be added tag is established and has a useful purpose
>   - I guess that a tag would still have to have a defined meaning, but does "useful purpose" apply to SCEE? Can a quest exist that adds a tag that has a defined meaning, but where no major app exists that uses it (yet)? 

**Yes**, tag should be clearly documented on wiki.osm.org (i.e. not disputed/confusing) and have at least some uses already (need not be much, but _more than one user_ adding the tags would be highly preferable). There is IMHO no need that there exists an app which uses the tag already (to avoid chicken and the egg problem).

> - Any answer the user can give must have an equivalent tagging (Quest should not reappear to other users when solved by one)
>   - Does this apply to SCEE? iirc, quests in SCEE can be designed around "Sometimes, there is no answer and the user should just select "hide"" 

I'd say **yes**. 
Note that users can hide quests in regular StreetComplete (although with more clicks - `Other answers` / `Can't say` / `Hide`). But the core of the issue is that OSM (and SC/SCEE by extension) is community effort. Hiding the quest however only ever hides the issue for **one** person (even less - only for one device of one person) - and the issue remains for everyone else, so it ends up being selfish towards other contributors.  

So, if there is more then one person interested in some tag (which one certainly should hope so!), there should be a way to eliminate a quest for for all users; and that requires a tag on the object. That requirement should not be such a big deal usually, though. After all, if one is [inventing](https://wiki.openstreetmap.org/wiki/ATYL) e.g. tag `foobar`, them may as well invent `foobar:signed=no` or similar tag too, to indicate that `foobar` cannot be found on that object.

> - Easily answerable by any pedestrian from the outside but a survey is necessary
>   - I guess this applys, maybe minus the "easily by any pedestrian" part 

**Yes**, it is IMHO OK to have quests which are harder to answer in SCEE, as well as those which can only be solved from a moving car or a train etc.

> - Not an overwhelming percentage of quests have the same answer (No spam)
>   - Does this apply? 

Probably not. When quest is disabled by default, user who willingly enables it despite it being spammy, they will only affect themselves. As SCEE does not try to cater to StreetComplete typical "total newbie who is easily turned away by such spam", we might as well accommodate mappers willing to obsess over some rare feature. I assume we'd put appropriate `defaultDisabledMessage` on those (e.g. `R.string.default_disabled_msg_difficult_and_time_consuming` or similar) 

So it might be **OK** to have even such quests.

> - Applies to a reasonable number of map data (Worth the effort)
>   - I think it applys, but would it still if I would implement the quests myself?

As above, I'm probably **OK** if it is not a hard requirement in case when proponent writes and tests PR themselves, even if very small number of users ends up enabling and using that quest.

Apart from of writing and testing the quest, only other major drawback is potential maintenance on major upgrades in more complex cases, but if turns out to be too much effort and nobody wants to dedicate time to fix it, it can always be removed, so that is not a big problem. Only other smaller problem is that it increases the list of quests and thus time needed to setup lists of presets, but that is the issue anyway, and one more isn't (hopefully; but see below!) going to make much of change.

> My second, intertwined, question is how I would go about "proposing" a large number of quests. 

This somewhat depends on _how large_ is that "large number of quests"? 10? 100? 1000? 

Some things which are not an issue for smaller "large" numbers (e.g. quests that only you or few other people end up using, like mentioned above, is likely OK for 10 new quests), suddenly become an issue on large "large" numbers (e.g. adding 1000 new quests that only you would end up using would almost certainly not be acceptable, as it would hugely complicate quest selection for everybody else without adding benefit to anybody but you)

> Should I open an issue for every single one and state that I would implement it myself? Should I just open pull requests? Should I open a discussion thread for all my ideas (one for every single one)? 

Well, if it is trivial quest and documented and established tag used in the wild, I think you could jump in directly with PR. 

However, especially for [ATYL](https://wiki.openstreetmap.org/wiki/ATYL) and/or rarely used tags, or some which have ambiguities or unknowns (e.g. how to tag every offered action in SCEE), I'd **recommend opening an issue first** - that way potential blockers as well comments which might influence implementation could be gathered before time is wasted on writing implementation which might need to be thrown away. And you'd gather thumbs up/down so you can see that some quests are more wanted than others, so you can prioritize working on them while trying to work out how to improve problematic ones.

And if how the tag is to be used is not crystal-clear, I'd recommend opening discussion on https://community.openstreetmap.org/c/general/tagging/70 (and linking that discussion to an GitHub issue about implementing that quest in SCEE), instead of using GitHub discussions (which would reach much smaller audience and get much less attention, and thus produce less useful results, if any)

> Should I start by opening and implementing one, and only after that is done another?

Yes, that would be a good idea if you have many proposals. Need not be strictly serialized on _one_ quest, but I would recommend keeping the number low, like 2-3 open quests that you work at once. That should work well to reducing your `idle` time (while people review/comment your one PR you can work on other one), while not being too overwhelming to others. It also reduces the risk that your enthusiasm gets spent in the middle, and we end up with 100 opened but never finished PRs. Also, humans are notoriously bad at multitasking, even (or especially) when they think they're not :smiley: 

Of course, if some the quests you start to write get stalled (e.g. while you wait for community discussions to gather), you can open and work on new ones until consensus is reached on older ones (if you wish).

**Comment by wielandb** [comment]
Thanks for your answers!

> Probably not. When quest is disabled by default, user who willingly enables it despite it being spammy, they will only affect themselves. As SCEE does not try to cater to StreetComplete typical "total newbie who is easily turned away by such spam", we might as well accommodate mappers willing to obsess over some rare feature. I assume we'd put appropriate defaultDisabledMessage on those (e.g. R.string.default_disabled_msg_difficult_and_time_consuming or similar)

Interesting, I always thought of the "no spam rule" as not meaning the map should not be spammed with quests for the user but rather that the OSM database should not be spammed with tags. (Which is a philosophy I only subscribe to to a certain extend)

> This somewhat depends on how large is that "large number of quests"? 10? 100? 1000?

Don't worry! :D I only have 9 quests on my list so far, but I still thought it would be a lot all at once.

> Yes, that would be a good idea if you have many proposals. Need not be strictly serialized on one quest, but I would recommend keeping the number low, like 2-3 open quests that you work at once. That should work well to reducing your idle time (while people review/comment your one PR you can work on other one), while not being too overwhelming to others. It also reduces the risk that your enthusiasm gets spent in the middle, and we end up with 100 opened but never finished PRs. Also, humans are notoriously bad at multitasking, even (or especially) when they think they're not 😃
> Of course, if some the quests you start to write get stalled (e.g. while you wait for community discussions to gather), you can open and work on new ones until consensus is reached on older ones (if you wish).

Understood! So I guess I'll just start with one of the more trvial ones as a PR, and for the ones where I would want some feedback first, I'll open an issue when the time has come. I will progress from simpler ones to complexer ones as time goes on.

**Reply by mnalis** [reply]
> Interesting, I always thought of the "no spam rule" as not meaning the map should not be spammed with quests for the user but rather that the OSM database should not be spammed with tags. (Which is a philosophy I only subscribe to to a certain extend)

Here is what upstream [Quest guidelines](https://github.com/streetcomplete/StreetComplete/blob/master/QUEST_GUIDELINES.md) say (emphasis mine):

> 💤 No spam: It must be possible to determine whether the quest should reasonably be asked. A quest which is to be answered in 99% of the time with the same answer is not a good quest, **don't bore the users**. I.e. asking for any road if it is a one-way road, is silly, because the vast majority of all roads will not be one-ways.

Of course, those quest guidelines are (as the name implies) primarily about what would make a good **quest in StreetComplete** (as opposed to for example Overlay, for which there exist [Overlay guidelines](https://github.com/streetcomplete/StreetComplete/blob/master/OVERLAY_GUIDELINES.md), or some other editor like EveryDoor). 
But, of course (as you note) in addition to any and all StreetComplete-specific Quest (and Overlay) guidelines, general [OSM good practices](https://wiki.openstreetmap.org/wiki/Good_practice) apply to _any_ editor, including StreetComplete (and SCEE).

So, if some specific tag/value is consider problematic for OSM database in vast majority of the cases (no matter which editor adds it!) then it should be documented as such on [wiki.osm.org](https://wiki.openstreetmap.org/), with clearly labeled exceptions where it is actually OK to use it (e.g. [highway=no](https://wiki.openstreetmap.org/wiki/Tag:highway=no) is not something one would go around adding to every object which is not a highway, but it should be used in few very specific cases only)

> Understood! So I guess I'll just start with one of the more trvial ones as a PR, and for the ones where I would want some feedback first, I'll open an issue when the time has come. I will progress from simpler ones to complexer ones as time goes on.

Sounds good to me! Helium314 would hopefully give his opinion in cases where he disagrees with my views eventually (but I'm not pinging him here as he's busy currently AFAIK), but hopefully they won't be in striking mismatch with my views :smiley_cat:

**Reply by Helium314** [reply]
Thanks @mnalis for extensively answering all of this.
I agree with your opinions and proposed procedure for issues / PRs, just some small additions:
> * Any answer the user can give must have an equivalent tagging (Quest should not reappear to other users when solved by one)
>   * Does this apply to SCEE? iirc, quests in SCEE can be designed around "Sometimes, there is no answer and the user should just select "hide""

It does not strictly apply, but as noted by @mnalis of course it would be best to keep it to rare cases instead of having many users wasting time by trying to answer the same unanswerable quest.

> * Not an overwhelming percentage of quests have the same answer (No spam)
>   * Does this apply?

See the (slightly) modified [SCEE quest guidelines](https://github.com/Helium314/SCEE/blob/modified/QUEST_GUIDELINES.md): "Not fulfilling any of the guidelines perfectly acceptable for SCEE, with the exception of the no spam guideline in case it leads to undesired tag spam.". What I mean here is e.g. the mentioned addition of `highway=no` to almost every object.
